arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/NonEmptySet.kt

```diff
@@ -0,0 +1,31 @@
+package arrow.core
+
+public class NonEmptySet<out T> private constructor(
+  private val elements: Set<T>
+) : Set<T> by elements {
+
+  public constructor(first: T, rest: Set<T>): this(setOf(first) + rest)
+
+  public operator fun plus(set: Set<@UnsafeVariance T>): NonEmptySet<T> = NonEmptySet(elements + set)
+
+  public operator fun plus(element: @UnsafeVariance T): NonEmptySet<T> = NonEmptySet(elements + element)
+
+  override fun toString(): String = "NonEmptySet(${this.joinToString()})"
+
+  override fun equals(other: Any?): Boolean {
+    return elements == other
+  }
```

Comment:
Shouldn't we account for types here?

```suggestion
  override fun equals(other: Any?): Boolean = when(other) {
    is Set<*> -> elements == other
    else -> false
  }
```

Reply:
You got it!